src/openssl.c

```diff
@@ -3403,11 +3403,26 @@ static BIO *getbio(lua_State *L) {
 } /* getbio() */
 
 
+static int pem_pw_cb(char *buf, int size, int rwflag, void *u) {
+	lua_State *L = (lua_State *) u;
+
+	if (lua_isnil(L, -1) || (lua_isfunction(L, -1) && lua_pcall(L, 0, 1, 0)))
+		return 0;
+
+	const char *pass = lua_tostring(L, -1);
+	if (!pass)
+		return 0;
+
+	strncpy(buf, pass, size);
+	return MIN(strlen(pass), (unsigned int) size);
```

Comment:
Make sure you pop the return value (also in the `!pass` case)

Reply:
I'm not sure why that would be necessary. The return value is removed from the stack when the parent Lua function (pk_new or pk_getPrivateKey) returns.

Reply:
Although at the moment you don't *have* to, it makes the code much easier to maintain if the stack is balanced.
But also, once you move the pkey allocation back again, it will give you a stable negative stack index.

Reply:
Please note that the callback will be invoked only if the private key is encrypted. Hence, it would not be safe to pop the return value inside the callback. The latest patch set pops the top value in pk_new after PEM reading, regardless of whether it is the password string, callback function, or error object.

Reply:
> Please note that the callback will be invoked only if the private key is encrypted.  Hence, it would not be safe to pop the return value inside the callback.

Why not? --> if you get past the `pcall`, then (after the `strncpy`), always pop.

> The latest patch set pops the top value in pk_new after PEM reading, regardless of whether it is the password string, callback function, or error object.

It may need to pop 2 values if both callbacks ran.

Reply:
The point is to avoid executing the callback function more than once as it may e.g. produce an interactive prompt to the user. A successful execution of lua_pcall pops the callback function from the stack, replacing it with the return value, i.e. the password string. If pem_pw_cb is invoked again, the string will be readily on top of the stack. Then lua_pcall will not be invoked but the string is read directly from the stack. The password will be popped once it is known that pem_pw_cb will no longer be called.